core/controllers/editor.py

```diff
@@ -613,59 +613,83 @@ class ImageUploadHandler(EditorHandler):
     # The string to prefix to the filename (before tacking the whole thing on
     # to the end of 'assets/').
     _FILENAME_PREFIX = 'image'
+    _decorator = None
 
-    @acl_decorators.can_edit_exploration
-    def post(self, exploration_id):
+    def post(self, entity_type, entity_id):
         """Saves an image uploaded by a content creator."""
 
-        raw = self.request.get('image')
-        filename = self.payload.get('filename')
-        if not raw:
-            raise self.InvalidInputException('No image supplied')
-
-        allowed_formats = ', '.join(
-            feconf.ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS.keys())
-
-        # Verify that the data is recognized as an image.
-        file_format = imghdr.what(None, h=raw)
-        if file_format not in feconf.ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS:
-            raise self.InvalidInputException('Image not recognized')
-
-        # Verify that the file type matches the supplied extension.
-        if not filename:
-            raise self.InvalidInputException('No filename supplied')
-        if filename.rfind('.') == 0:
-            raise self.InvalidInputException('Invalid filename')
-        if '/' in filename or '..' in filename:
-            raise self.InvalidInputException(
-                'Filenames should not include slashes (/) or consecutive dot '
-                'characters.')
-        if '.' not in filename:
-            raise self.InvalidInputException(
-                'Image filename with no extension: it should have '
-                'one of the following extensions: %s.' % allowed_formats)
+        if entity_type == fs_domain.ENTITY_TYPE_EXPLORATION:
+            _decorator = acl_decorators.can_edit_exploration
+        elif entity_type == fs_domain.ENTITY_TYPE_TOPIC:
+            _decorator = acl_decorators.can_edit_topic
+        elif entity_type == fs_domain.ENTITY_TYPE_SKILL:
+            _decorator = acl_decorators.can_edit_skill
+        elif entity_type == fs_domain.ENTITY_TYPE_STORY:
+            _decorator = acl_decorators.can_edit_story
+
+        @_decorator
+        def _upload_image(self, entity_id, entity_type):
```

Comment:
Swap entity_type and entity_id

Reply:
Similar comment, as above. (Even @anubhavsinha98 has the same issue). Since I am using existing decorators - can_edit_exploration, can_edit_topic etc - which require the first argument to be the ID of the entity, this function (on which those decorators act on now), has to have the arguments in the correct order.
If I do swap them, I don't think we can reuse those decorators unless we change those decorators themselves which will affect all the other controllers that use them.

Reply:
How about something like this? (You can copy the following into a python script and run it; it's self-contained. Tests are at the bottom)

Basically, make a more general can_edit_entity decorator that delegates to the existing ones, and use that more general decorator here. Then you don't need all the if/else above.

```
class ExpectedException(Exception):
    pass

def can_edit_question(handler):

    def test_can_edit_question(self, question_id):
        if question_id == 'invalid':
            raise ExpectedException('Invalid question ID: %s' % question_id)
        return handler(self, question_id)

    test_can_edit_question.__wrapped__ = True

    return test_can_edit_question


def can_edit_entity(handler):

    def test_can_edit_entity(self, entity_type, entity_id):
        if entity_type == 'question':
            # This raises an error if the question checks fail.
            can_edit_question(handler)
        else:
            raise ExpectedException('Invalid entity type')

        return handler(self, entity_type, entity_id)

    test_can_edit_entity.__wrapped__ = True

    return test_can_edit_entity



class QuestionHandler(object):
    @can_edit_question
    def get(self, question_id):
        return 'Retrieved question %s' % question_id


class EntityHandler(object):
    @can_edit_entity
    def get(self, entity_type, entity_id):
        return 'Retrieved entity %s:%s' % (entity_type, entity_id)


assert QuestionHandler().get('abc') == 'Retrieved question abc'
assert EntityHandler().get('question', 'abc') == 'Retrieved entity question:abc'

try:
    QuestionHandler().get('invalid')
except ExpectedException as e:
    assert str(e) == 'Invalid question ID: invalid'

try:
    EntityHandler().get('question', 'invalid')
except ExpectedException as e:
    assert str(e) == 'Invalid question ID: invalid'

try:
    EntityHandler().get('invalid_type', 'abc')
except ExpectedException as e:
    assert str(e) == 'Invalid entity type'
```

Reply:
Oops, sorry, please ignore the previous comment -- that test snippet was buggy. Just because one writes try/except doesn't mean that an exception is actually raised :)

Here is a better one:

```
import functools
import unittest

class ExpectedException(Exception):
    pass

def can_edit_question(handler):

    def test_can_edit_question(self, question_id):
        if question_id == 'invalid':
            raise ExpectedException('Invalid question ID: %s' % question_id)
        return handler(self, question_id)

    test_can_edit_question.__wrapped__ = True

    return test_can_edit_question


def can_edit_entity(handler):

    def test_can_edit_entity(self, entity_type, entity_id):
        if entity_type == 'question':
            # This creates a modified handler function that has the correct
            # signature for can_edit_question().
            modified_handler = functools.partial(handler, 'question')
            # This raises an error if the question checks fail.
            can_edit_question(modified_handler)(self, entity_id)
        else:
            raise ExpectedException('Invalid entity type')

        return handler(self, entity_type, entity_id)

    test_can_edit_entity.__wrapped__ = True

    return test_can_edit_entity



class QuestionHandler(object):
    @can_edit_question
    def get(self, question_id):
        return 'Retrieved question %s' % question_id


class EntityHandler(object):
    @can_edit_entity
    def get(self, entity_type, entity_id):
        return 'Retrieved entity %s:%s' % (entity_type, entity_id)


class TestCase(unittest.TestCase):
    def test(self):
        self.assertEqual(
            QuestionHandler().get('abc'), 'Retrieved question abc')
        self.assertEqual(
            EntityHandler().get('question', 'abc'),
            'Retrieved entity question:abc')

        with self.assertRaisesRegexp(
                ExpectedException, 'Invalid question ID: invalid'):
            QuestionHandler().get('invalid')

        with self.assertRaisesRegexp(
                ExpectedException, 'Invalid question ID: invalid'):
            EntityHandler().get('question', 'invalid')

        with self.assertRaisesRegexp(
                ExpectedException, 'Invalid entity type'):
            EntityHandler().get('invalid_type', 'abc')

if __name__ == '__main__':
    unittest.main()
```

Thanks to @anubhavsinha98 for pointing out that there was a flaw in the previous code.

Reply:
Done